### ISSUE TITLE
What is the right performant C++ interface for online prediction?

### ISSUE THREAD

**Issue by WhoisZihan** [dmlc/xgboost#7150]
Hi, I'd like to start serving xgboost model for online prediction in C++, but I am not sure which interface is fastest, and would appreciate any suggestion.

The [XGBoosterPredict](https://xgboost.readthedocs.io/en/stable/dev/c__api_8h.html#adc14afaedd5f1add105d18942a4de33c) is marked deprecated, pointing me to [XGBoosterPredictFromDMatrix](https://xgboost.readthedocs.io/en/stable/dev/c__api_8h.html#a843f60933367fb27e78dc49bdbbe9f32) instead, but the latter seems not stable yet based on issue #6632 (its implementation differs in 1.4.2 and master). I also find an interface called [InplacePredictImpl](https://github.com/dmlc/xgboost/blob/bd1f3a38f03e8131a4eca9890bb98851f0751c1a/src/c_api/c_api.cc#L750) that sounds the fastest by its name, but have no idea how to use it.

So, to figure it out, I manually trained a model using the [rank demo in the source code](https://github.com/dmlc/xgboost/tree/master/demo/rank), and the prediction input is from test matrix(in csr format) without any modification. Here is the experiment I perform using 1.4.2.

(1) `XGBoosterPredict`
The config I set for `XGBoosterPredict` is
```c++
s = XGBoosterPredict(handle, dinput, /*option_mask*/ 0, /*ntree_limit*/ 4, /*training*/ 0, &out_len, &out_result);
```

Running it 10000 rounds takes 1275806us, which is 127.58us per round

(2) `XGBoosterPredictFromDMatrix`
```c++
s = XGBoosterPredictFromDMatrix(handle, dinput, R"({"type": 0, "ntree_limit":4,"iteration_begin":0, "iteration_end":4,"training":false,"strict_shape":false})", &out_shape, &out_dim, &out_result);
```

Suprisingly, running it 10000 rounds takes only 54608 us, which is 5.46us per round! **I believe the above config is exactly the same with the config in (1), and they both seem to call `learner->Predict` internally, so I am not sure why it is so much faster**. Am I missing anything?

P.S. I only measure the time consume by predict method themselves, not including the feature preparation phase.

(3) `InplacePredictImpl`

I did not try this one because I am not sure how to correctly construct the input yet.

<br></br>
So my confusions are
- Am I using `XGBoosterPredictFromDMatrix` correctly? Is is really extremely faster than `XGBoosterPredict`?
- Can we avoid the json config in the future? I doubt it would not be performance friendly when the request is from a rpc request. Although the config is usually small, but parsing the rpc bytes + formatting the json config string + parsing the json string again inside the function, seems a lot extra work that could have been avoided. Dealing with json means introducing extra json library and larger codebase. So if you plan to avoid json config in the future, I'd be really happy about it :)
-  Is there any document about how to use `InplacePredictImpl` or `XGBoosterPredictFromDMatrix`?

Or simply speaking, as the title suggests, what is the most performant C++ interface for online prediction? I would like to step on the right road from the beginning, so that I don't have to refactor my code later on.

Thanks!

**Comment by trivialfis** [comment]
Before answering questions, there are specialized tools for running inference servers, model deployment is becoming its own field now.  To get started you probably want to look at https://github.com/dmlc/treelite and FIL https://medium.com/rapids-ai/rapids-forest-inference-library-prediction-at-100-million-rows-per-second-19558890bc35 .  @hcho3 has better insight into this field.  (full disclosure, we both work for NVIDIA).

> but the latter seems not stable yet based on issue #6632 (its implementation differs in 1.4.2 and master).

The RFC describes a general direction (being flexible and feature-complete at C level).  For the implementation details, we decided to move from using C struct to JSON for parameters due to the difficulty of handling FFI.  So the details in RFC are outdated by now.

> Am I using XGBoosterPredictFromDMatrix correctly?

For both prediction functions, you should specify only the `iteration_begin/end`, `ntree_limit` should not be used after 1.4.

> Can we avoid the JSON config in the future?

Probably no.  XGBoost accepts a lot of parameters and we want to keep stability for our API functions.  JSON is used for its flexibility.  With it, we can add or remove parameters without making breaking changes.  For example, both the `training` parameter and `iteration_begin/end` parameter are new, and we don't want to complicate the prediction function too much since XGBoost is not designed to handle model deployment.

> Is there any document about how to use InplacePredictImpl or XGBoosterPredictFromDMatrix

For C API, the doc in the header is the only resource right now.  It should be faster considering that it doesn't need caching nor the construction of DMatrix.  I can write a demo if needed.

> what is the most performant C++ interface for online prediction?

Inplace prediction.  You can pass the raw data directly into XGBoost without making any copy.  Also, XGBoost doesn't have to perform caching.

**Comment by trivialfis** [comment]
Aside from the above comment, JSON is not that slow for a small parameter object, we actually verify its correctness on floating-point conversion by enumerating all 32-bit floating-point.  For online prediction, it might have some overhead, but it's efficient enough for mini-batch prediction.

**Comment by trivialfis** [comment]
Added a demo.

**Comment by WhoisZihan** [comment]
> Before answering questions, there are specialized tools for running inference servers, model deployment is becoming its own field now. To get started you probably want to look at https://github.com/dmlc/treelite and FIL https://medium.com/rapids-ai/rapids-forest-inference-library-prediction-at-100-million-rows-per-second-19558890bc35 . @hcho3 has better insight into this field. (full disclosure, we both work for NVIDIA).

Thanks for the reference! I'll take a look at them.

> > but the latter seems not stable yet based on issue #6632 (its implementation differs in 1.4.2 and master).
> 
> The RFC describes a general direction (being flexible and feature-complete at C level). For the implementation details, we decided to move from using C struct to JSON for parameters due to the difficulty of handling FFI. So the details in RFC are outdated by now.
> 
> > Am I using XGBoosterPredictFromDMatrix correctly?
> 
> For both prediction functions, you should specify only the `iteration_begin/end`, `ntree_limit` should not be used after 1.4. I don't expect such a huge difference in performance, might have to inspect your code to decide whether you are accidentally using the prediction cache.

I didn't realize there was prediction cache, but it sounds reasonable for the speedup. I did use the same input for the whole 10000 rounds within a simple for loop. Also I skipped the 1st round for the warmup(as we usually do for tensorflow models).

I'll check it later to see if I walked into the cached path. Thanks for the hint.

> > Can we avoid the JSON config in the future?
> 
> Probably no. XGBoost accepts a lot of parameters and we want to keep stability for our API functions. JSON is used for its flexibility. With it, we can add or remove parameters without making breaking changes. For example, both the `training` parameter and `iteration_begin/end` parameter are new, and we don't want to complicate the prediction function too much since XGBoost is not designed to handle model deployment.

Flexibility makes sense.

> > Is there any document about how to use InplacePredictImpl or XGBoosterPredictFromDMatrix
> 
> For C API, the doc in the header is the only resource right now. It should be faster considering that it doesn't need caching nor the construction of DMatrix. I can write a demo if needed.

That would be great! I'll look into the C API demo.

> > what is the most performant C++ interface for online prediction?
> 
> Inplace prediction. You can pass the raw data directly into XGBoost without making any copies. Also, XGBoost doesn't have to perform caching.
